vllm/v1/core/kv_cache_manager.py

```diff
@@ -355,43 +305,72 @@ def _touch(self, blocks: List[KVCacheBlock]) -> None:
-            token_ids: A list of token IDs to add.
-            parent_block: The parent block. None if this is the
-                first block.
+    @staticmethod
+    def _cache_full_blocks(
+        request: Request,
+        blk_start_idx: int,  # TODO: use prev_block's info instead.
+        full_blocks: List[KVCacheBlock],
+        prev_block: Optional[KVCacheBlock],
+        cached_block_hash_to_block: Dict[BlockHashType, Dict[int,
+                                                             KVCacheBlock]],
+        block_size: int,
+    ) -> None:
+        """Cache a list of full blocks for prefix caching.
 
-        Returns:
-            The starting token ID index that has not been added to the blocks
-            due to insufficient given blocks.
+        This function takes a list of blocks that will have their block hash
+        metadata to be updated and cached. Given a request, it computes the
+        block hashes for the blocks starting from `blk_start_idx` to the end
+        of the request's full blocks, updating the metadata for each block
+        and caching them in the `cached_block_hash_to_block`.
+
+        Args:
+            request: The request to cache the blocks.
+            blk_start_idx: The index of the first block in the request's blocks
+                to cache.
+            full_blocks: The list of blocks to update hash metadata.
+            prev_block: The previous block in the chain.
+            cached_block_hash_to_block: The cache of block hashes.
+            block_size: The size of a block.
         """
-        token_id_start = 0
-        for curr_block in blocks:
-            # If all token IDs are added, then the rest of the blocks are
-            # preallocated blocks, so we only need to update the
-            # parent_block_id. FIXME
-            if token_id_start == len(token_ids):
-                continue
-
-            # Add token IDs to the empty slots in the block.
-            empty_slots = self.block_size - len(curr_block.token_ids)
-            token_id_end = min(token_id_start + empty_slots, len(token_ids))
-            curr_block.token_ids.extend(token_ids[token_id_start:token_id_end])
-            # Cache the block if it becomes full.
-            if len(curr_block.token_ids) == self.block_size:
-                self._cache_full_block(curr_block, parent_block)
-            parent_block = curr_block
-            token_id_start = token_id_end
-        return token_id_start
+
+        # Some of the full blocks may be computed blocks cached previously.
+        # Remove them from the cache.
+        for blk in full_blocks:
+            blk.reset_hash_metadata()
+
+            if blk.block_hash in cached_block_hash_to_block:
+                del cached_block_hash_to_block[blk.block_hash][blk.block_id]
+
+                if len(cached_block_hash_to_block[blk.block_hash]) == 0:
+                    del cached_block_hash_to_block[blk.block_hash]
```

Comment:
This logic shouldn't be here IIUC, because it's not guaranteed that you will always call `_cache_full_blocks()` for all blocks you get from `_get_new_blocks()` (if they are not full yet).

Reply:
Oh yeah, i realized this too. Good catch.

Reply:
Added a test for this.